docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,74 @@
+
+## Proposed Solution
+
+The Connection Diagnostics will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionDiagnostics::State`. The Connection Diagnostics will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful. 
+
+In the event of a probe failure, the Connection Diagnostics will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted and the Connection Diagnostics will enter the Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Diagnostics.
+
+> For specific details about how these failures are handled and communicated with the user, please refer to this [Figma document](https://www.figma.com/file/GZKigVZbnCisMJNU3Hc9xM/Connection-error-messaging?type=design&node-id=26-215&mode=design&t=kfYnUKuXQ9nsGDRR-0).
+
+## Implementation Plan
+
+```c++
+enum ConnectionDiagnosticsStates {
+  StateProbingInternet,
+  StateCheckingFirewall,
+  StateProbingServerLocation,
+  StateCheckingSubscription,
+  StateIdle,
+}
+```
+
+### Upon VPN Activation (`StateOff` to `StateConnecting`) - Phase 1
+Perform a single cycle going through all the Connection Diagnostics probes.
+- If a probe fails:
+  - Set Controller to `StateOff`
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+  - VPN will remain in `StateOff`
+  - Connection Diagnostics will go into `StateIdle`
+- If all probes pass:
+  - Activate VPN (Controller `StateOn`)
+  - Connection Diagnostics will go into `StateIdle`
+
+### VPN is already active (`StateOn`) - Phase 2
+If the VPN is already active and Connection Health stability goes into no signal, initiate a diagnostics cycle to perform a set of checks to identify the underlying cause.
+
+- If a probe fails:
```

Comment:
If we hit NoSignal state and run diagnostic checks, we'd expect that a probe would fail, right? Is there any corner case possiblity to hit NoSignal but end up with all successful checks? Even if we can't consider a way this could happen, what would the user see? Just a NoSignal red state with no modal? If that happens, we should at the very least log the heck out of it (and maybe send a Glean event), no?

Reply:
> Is there any corner case possiblity to hit NoSignal but end up with all successful checks? Even if we can't consider a way this could happen, what would the user see? Just a NoSignal red state with no modal?

Yes I'm adding a little more clarification on this to the tech spec but as you mentioned, if all probes pass and we are still in No Signal, then we won't show anything beyond "No Signal" in the controller

```
- If all probes pass but the stability is still `NoSignal`:
  - Something else has gone wrong and are not able to determine the cause. VPN continues to stay in No Signal but we will not show any further modals to the user
  - This will be recorded in the logs for debugging purposes
```

Reply:
@mcleinman I agree with recording the frequency of how often we end up in No Signal without a specific diagnostic. I don't think we should be getting into the specifics of what metrics to record here though, right? I assume that would be a Miro doc further down the line?

Reply:
I don't feel like it *needs* to be acknowledged here, but I want to make sure we don't forget it. If the eventual metrics doc is not engineer-led, I'm not sure this situation would be remembered. But I don't think you need to add something here.